quickwit/quickwit-indexing/src/source/queue_sources/message.rs

```diff
@@ -0,0 +1,295 @@
+#[derive(Debug, Clone, PartialEq, Eq)]
+pub struct MessageMetadata {
+    /// The handle that should be used to acknowledge the message or change its visibility deadline
+    pub ack_id: String,
+
+    /// The unique message id assigned by the queue
+    pub message_id: String,
+
+    /// The approximate number of times the message was delivered. 1 means it is
+    /// the first time this message is being delivered.
+    pub delivery_attempts: usize,
+
+    /// The first deadline when the message is received. It can be extended later using the ack_id.
+    pub initial_deadline: Instant,
+}
+
+/// The raw messages as received from the queue abstraction
+pub struct RawMessage {
+    pub metadata: MessageMetadata,
+    pub payload: OwnedBytes,
+}
+
+impl fmt::Debug for RawMessage {
+    fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
+        f.debug_struct("RawMessage")
+            .field("metadata", &self.metadata)
+            .field("payload", &"<bytes>")
+            .finish()
+    }
+}
+
+#[derive(Error, Debug)]
+pub enum PreProcessingError {
+    /// The message can be acknowledged without processing
+    #[error("skippable message: {reason}")]
+    Discardable {
+        reason: &'static str,
+        ack_id: String,
+    },
+    #[error("unexpected message format: {0}")]
+    UnexpectedFormat(#[from] anyhow::Error),
+}
+
+impl RawMessage {
+    pub fn pre_process(
+        self,
+        message_type: MessageType,
+    ) -> Result<PreProcessedMessage, PreProcessingError> {
+        let payload = match message_type {
+            MessageType::S3Notification => PreProcessedPayload::ObjectUri(
+                uri_from_s3_notification(&self.payload, &self.metadata.ack_id)?,
+            ),
+            MessageType::RawUri => {
+                let payload_str = read_to_string(self.payload).context("failed to read payload")?;
+                PreProcessedPayload::ObjectUri(Uri::from_str(&payload_str)?)
+            }
+        };
+        Ok(PreProcessedMessage {
+            metadata: self.metadata,
+            payload,
+        })
+    }
+}
+
+#[derive(Debug, PartialEq, Eq)]
+pub enum PreProcessedPayload {
+    ObjectUri(Uri),
+}
+
+impl PreProcessedPayload {
+    pub fn partition_id(&self) -> PartitionId {
+        match &self {
+            Self::ObjectUri(uri) => PartitionId::from(uri.as_str()),
+        }
+    }
+}
+
+/// A message that went through the minimal transformation to discover its
+/// partition id. Indeed, the message might be discarded if the partition was
+/// already processed, so it's better to avoid doing unnecessary work at this
+/// stage.
+#[derive(Debug, PartialEq, Eq)]
+pub struct PreProcessedMessage {
+    pub metadata: MessageMetadata,
+    pub payload: PreProcessedPayload,
+}
+
+impl PreProcessedMessage {
+    pub fn partition_id(&self) -> PartitionId {
+        self.payload.partition_id()
+    }
+}
+
+fn uri_from_s3_notification(message: &OwnedBytes, ack_id: &str) -> Result<Uri, PreProcessingError> {
```

Comment:
```suggestion
fn uri_from_s3_notification(message: &[u8], ack_id: &str) -> Result<Uri, PreProcessingError> {
```

Reply:
> Taking &OwnedBytes is never a good idea 

`&[u8]` is much better here, thanks

> it is possible for a notification to not be about a PUT (I think?). In that case, I don't think this is a great idea to return it as an error
We cannot really do much more on the caller's side by knowing that the message has a wrong event type vs. any other kind of malformation, so I'll leave the return type as is for now.

Reply:
I don't understand your point.